core/iwasm/libraries/lib-wasi-threads/test/sync_primitives.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2023 Amazon.com Inc. or its affiliates. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include <stdbool.h>
+
+/* Mutex */
+
+typedef int mutex_t;
+
+void
+mutex_init(mutex_t *mutex)
+{
+    *mutex = 0;
+}
+
+static bool
+try_mutex_lock(mutex_t *mutex)
+{
+    int expected = 0;
+    return __atomic_compare_exchange_n(mutex, &expected, 1, false,
+                                       __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+}
+
+void
+mutex_lock(mutex_t *mutex)
+{
+    while (!try_mutex_lock(mutex))
+        __builtin_wasm_memory_atomic_wait32(mutex, 1, -1);
+}
+
+void
+mutex_unlock(mutex_t *mutex)
+{
+    __atomic_store_n(mutex, 0, __ATOMIC_SEQ_CST);
+    __builtin_wasm_memory_atomic_notify(mutex, 1);
+}
+
+/* Barrier */
+
+typedef struct {
+    int count;
+    int num_threads;
+    int mutex;
+    int ready;
+} barrier_t;
+
+void
+barrier_init(barrier_t *barrier, int num_threads)
+{
+    barrier->count = 0;
+    barrier->num_threads = num_threads;
```

Comment:
why not just use one of these?

Reply:
Right, the point of having both is to reset the count to 0 once it reaches `num_threads`. I'll update the code.